docs(next-drupal): updated docs

MontiMarco92 opened the pull request:
This pull request is for: (mark with an "x")

- [ ] `examples/*`
- [ ] `modules/next`
- [x] `packages/next-drupal`
- [ ] `starters/basic-starter`
- [ ] `starters/graphql-starter`
- [ ] `starters/pages-starter`
- [ ] Other

GitHub Issue: #691 

- [ ] I need help adding tests. (mark with an "x")
      _Code changes need test coverage. If you don't know
      how to make tests, check this box to ask for help._

## Describe your changes

I've updated the documentation to reflect the changes of the new 2.0 version. Basically I've updated all the pages related to the following sections on the `/docs/` route.

- Drupal Client
- Working with JSON:API
- Building Pages
- Customization
- Reference

![image](https://github.com/chapter-three/next-drupal/assets/88577526/49963516-1f90-4118-8ac0-7d3a7fc47d1f)

I followed the same structure and did not add extra sections (perhaps some other additional example).

On the `Reference` section I've removed the old methods and added some new ones. The complete list of the updated methods are:

- getResource
- getResourceByPath
- getResourceCollection
- createResource
- createFileResource
- updateResource
- deleteResource
- getResourceCollectionPathSegments
- translatePath
- constructPathFromSegment
- buildEndpoint (not sure if it's intended to be used by users or only internally)
- getAccessToken
- getMenu
- getView
- getSearchIndex
- buildUrl
- fetch
- deserialize
- getAuthorizationHeader


**Methods that have not been added:**
- fetchResourceEndpoint
- addLocalePrefix
- validateDraftUrl

Let me know if it's relevant that they are added as well

*Note*: On some of them, I assumed the `next` revalidation options proposed on this [PR](https://github.com/chapter-three/next-drupal/pull/784) will be accepted. However, it can be the case that this feature is modified and hence, the docs will have to be updated as well. Besides, not all of the methods have been extended to take this new `next` option as explained on the PR, since I'm waiting for feedback.

*Note 2*: I've left some HTML comments on some places just to note that there are some questions on the matter (at least from me). 

Hope this helps!

backlineint commented on www/content/docs/pages.mdx:
```diff
@@ -22,136 +30,94 @@ const node = await drupal.getResource(
 ## Dynamic pages
 
-You can use Next.js [dynamic route](https://nextjs.org/docs/basic-features/pages#pages-with-dynamic-routes) to build static pages for Drupal entity types.
+You can use Next.js [dynamic route](https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes) to build static pages for Drupal entity types.
```
```suggestion
You can use Next.js [dynamic routes](https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes) to build static pages for Drupal entity types.
```

backlineint commented on www/content/docs/pages.mdx:
```diff
@@ -168,15 +134,20 @@ export async function getStaticProps(context) {
   }
 
-  const node = await drupal.getResourceFromContext(type, context, {
+  const node = await drupal.getResourceByPath(path, {
```
Since we have already translated the path here, sing getResource could save us an API call (depending on how things are being cached)

```suggestion
  const node = await drupal.getResource(type, path.entity.uuid {
```

backlineint commented on www/content/docs/pages.mdx:
```diff
@@ -168,15 +134,20 @@ export async function getStaticProps(context) {
-    },
+  // Render different Components based on Node type
+  if (node.type === "node--page") {
+  return <PageComponent />
```
We'd have to pass some kind of props here (and also for the article below), if it was the entire node object it would be:

```suggestion
  return <PageComponent node={node} />
```

backlineint reviewed (commented):
Doing a review in phases here, because this is a big PR :) I'm through the root of docs, up to the reference section.

So far this looks great @MontiMarco92

There are a few things I think we'll have to flesh out beyond the changes here (although I may not have gotten to it yet in the PR):
* Upgrading from V1 to V2
* A home for NextDrupalPages specific content.
* Some additional detail related to caching and the app router.

But I'm thinking we should be able to get these changes merged as a first step.

Hope to have some time to continue the review soon...

backlineint reviewed (commented):
Sharing another round of review comments - about halfway through.

backlineint commented on www/content/tutorials/preview-mode/configure-preview-routes.mdx:
I'm not really sure of the best way to handle this, but technically this content will still be relevant to those using the pages router. Maybe this could be included as an additional section on the configure draft routes entry?

MontiMarco92 replied:
I haven't addressed this, cause I think It has to be made a decision regarding the next drupal pages specific documentation.

backlineint commented on www/content/tutorials/quick-start/index.mdx:
```diff
@@ -5,7 +5,7 @@ weight: 10
 ---
 
-This tutorial will help you get started with Next.js for Drupal using the [`DrupalClient`](/docs/client) and **JSON:API**.
+This tutorial will help you get started with Next.js for Drupal using the [`NextDrupal`](/docs/client) and **JSON:API**.
```
```suggestion
This tutorial will help you get started with Next.js for Drupal using the [`NextDrupal Client`](/docs/client) and **JSON:API**.
```

backlineint reviewed (changes requested):
Finished a full pass at review here. Excellent work @MontiMarco92 !

Next steps:
* Incorporate the requested revisions.
* Find a home for NextDrupalPages/Page Router specific content.
* Come to a conclusion about how we handle things related to caching and the app router - do we remove for now, or do we integrate https://github.com/chapter-three/next-drupal/pull/784 before we merge this.
* Add a section on upgrading from V1 to V2 (the changelog in the codebase probably gives us a start here)

Happy to have your continued help here @MontiMarco92, but don't expect you to cover all of this. As a result, we'll need to come up with some way to collaborate on this PR.

MontiMarco92 commented:
Thanks a lot for your review @backlineint and your feedback! I'll be happy to address whatever changes I can when I have time. 
- If by the time I come back on this, the conclusion on the caching issue #784 has not yet being defined I will remove all the things that reference that on this PR.
- I might further ask your opinion/directive on how/where to locate the NextDrupal pages router docs.

Thanks again and I hope this will help

MontiMarco92 commented on www/content/tutorials/on-demand-revalidation/configure-entity-types.mdx:
```diff
@@ -5,6 +5,16 @@ weight: 350
+
+*  Comment: The revalidator plugin used to work out of the box with path-based revalidation. However, Next.js implemented Tag-based revalidation, which is
+pretty useful for revalidating Menus on the FE whenever a Drupal Menu is updating. I'm just leaving this comment to note this, since I do not know
+if the new version of drupal module supports this out of the box.
```
Removed.

MontiMarco92 commented on www/content/docs/serializer.mdx:
file restored

MontiMarco92 commented on www/content/docs/reference/preview.mdx:
file restored

MontiMarco92 commented on www/content/docs/reference/getview.mdx:
````diff
@@ -57,6 +63,22 @@ const view = await drupal.getView("articles--promoted", {
+
+```ts
+const view = await drupal.getView("articles--promoted", {
+  next: { revalidate: 3600 }, // revalidate every hour`
````
Removed.

MontiMarco92 commented on www/content/docs/reference/translatepath.mdx:
```diff
@@ -8,18 +8,24 @@ const path = await drupal.translatePath(
+    next
   }
-): Promise<DrupalTranslatedPath>
+): Promise<DrupalTranslatedPath | null>
```
Removed

MontiMarco92 commented on www/content/docs/reference/translatepathfromcontext.mdx:
file restored

MontiMarco92 commented on www/content/docs/authentication.mdx:
```diff
@@ -192,13 +177,21 @@ const articles = await drupal.getResourceCollection("node--article", {
 
 If you're using [NextAuth](https://next-auth.js.org) you can use the results from `getSession`:
 
+<!--
```
removed

MontiMarco92 commented on www/content/docs/configuration.mdx:
```diff
@@ -123,6 +125,14 @@ You can find more info about using a custom fetcher [here](/docs/fetcher).
 
 ### cache
 
+<!--
```
Removed

MontiMarco92 commented on www/content/docs/configuration.mdx:
```diff
@@ -171,16 +181,13 @@ Set whether the client should use authenticated requests by default. If set to `
-
-- **Default value**: `false`
-- **Required**: No
+<!--
```
removed

MontiMarco92 commented on www/content/docs/fetcher.mdx:
```diff
@@ -1,9 +1,17 @@
+************************************************************
+*  TODO: Remove this block before publishing.
+
+*  Comment: Not sure how a custom fetcher will work because of the caching options of the Nextjs extended fetch API
```
This needs to be addressed. In order to use a custom fetcher (like Axios), you can indeed use this 'unstable_cache' api. However, in order to handle caching on this custom fetcher, revalidation props need to be supported. Therefore I reckon a decision regarding revalidation options must be made first

backlineint replied:
That makes sense. Depending on when the existing revalidation PRs are resolved we'll either update this section or could outline the known limitations here.

MontiMarco92 commented on www/content/docs/fetcher.mdx:
````diff
@@ -39,10 +47,7 @@ const customFetcher = (url, options) => {
+export const drupal = new NextDrupal(process.env.NEXT_PUBLIC_DRUPAL_BASE_URL, {
+  fetcher: customFetcher,
+})
 ```
````
See comment above

MontiMarco92 commented on www/content/docs/fetching-resources.mdx:
````diff
@@ -188,6 +169,9 @@ const article = await drupal.getResourceCollection(
 )
 ```
 
+<!--
````
removed

MontiMarco92 commented on www/content/docs/preview-mode.mdx:
file restored

MontiMarco92 commented on www/content/docs/reference/buildEndpoint.mdx:
fixed

MontiMarco92 commented on www/content/docs/reference/getentryforresourcetype.mdx:
file restored

MontiMarco92 commented on www/content/docs/reference/getmenu.mdx:
```diff
@@ -10,44 +10,51 @@ You need to install the [JSON:API Menu Items](https://www.drupal.org/project/jso
   - `defaultLocale: string`: The default locale of the site.
   - `withCache: boolean`: Set `withCache` if you want to store and retrieve the menu from cache.
   - `cacheKey: string`: The cache key to use.
+  - `next: NextFetchRequestConfig`
```
removed

MontiMarco92 commented on www/content/docs/reference/getmenu.mdx:
````diff
@@ -67,3 +74,19 @@ const menu = await drupal.getMenu("main", {
   cacheKey: "menu--main",
 })
 ```
+
````
removed

MontiMarco92 commented on www/content/docs/reference/getpathfromcontext.mdx:
file restored

MontiMarco92 commented on www/content/docs/reference/getresource.mdx:
```diff
@@ -27,14 +28,19 @@ const resource = await drupal.getResource<T = JsonApiResource>(
   - `defaultLocale: string`: The default locale of the site.
   - `withCache: boolean`: Set `withCache` if you want to store and retrieve the resource from cache.
   - `cacheKey: string`: The cache key to use.
+  - `next: NextFetchRequestConfig`
```
removed

MontiMarco92 commented on www/content/docs/reference/getresourcebypath.mdx:
```diff
@@ -13,6 +13,7 @@ const resource = await drupal.getResourceByPath<T = JsonApiResource>(
     locale,
     defaultLocale,
     isVersionable,
+    next,
```
removed

MontiMarco92 commented on www/content/docs/reference/getresourcebypath.mdx:
````diff
@@ -58,6 +64,31 @@ const { data, meta, links } = await drupal.getResourceByPath(
 )
 ```
 
+- Get the `es` translation for a page.
````
removed

MontiMarco92 commented on www/content/docs/reference/getresourcecollection.mdx:
```diff
@@ -12,6 +12,7 @@ const resource = await drupal.getResourceCollection<T = JsonApiResource[]>(
     deserialize,
     locale,
     defaultLocale,
+    next,
```
removed

MontiMarco92 commented on www/content/docs/reference/getresourcecollection.mdx:
````diff
@@ -57,6 +63,22 @@ const { data, meta, links } = await drupal.getResourceCollection("node--page", {
+
+```ts
+const articles = await drupal.getResourceCollection("node--article", {
+  next: { revalidate: 3600 }, // revalidate every hour
````
removed

MontiMarco92 commented on www/content/docs/reference/getresourcecollectionfromcontext.mdx:
file restored

MontiMarco92 commented on www/content/docs/reference/getresourcecollectionpathsegments.mdx:
```diff
@@ -0,0 +1,99 @@
+    pathPrefix,
+    locale,
+    defaultLocale,
+    next,
```
removed

MontiMarco92 commented on www/content/docs/reference/getresourcefromcontext.mdx:
file restored

MontiMarco92 commented on www/content/docs/reference/getsearch.mdx:
```diff
@@ -18,6 +18,7 @@ const resources = await drupal.getSearchIndex<T = JsonApiResource[]>(
     deserialize,
     locale,
     defaultLocale,
+    next
```
removed

MontiMarco92 commented on www/content/docs/reference/getstaticpathsfromcontext.mdx:
file restored

MontiMarco92 commented on www/content/docs/reference/getview.mdx:
```diff
@@ -18,6 +18,7 @@ const view = await drupal.getView<T = JsonApiResource>(
     deserialize,
     locale,
     defaultLocale,
+    next
```
removed

MontiMarco92 commented:
@backlineint I have addressed some of the requested modifications. I've left some comments as well.

- Regarding the previously removed files that are still valid for the `NextDrupalPages` client, I've restored them, but haven't referenced them on the menu. They should be referenced once you know where this client specific docs are going to be placed. 

- I've removed all the revalidation related references since a decision regarding the caching system #784  needs to be made.

backlineint reviewed (approved):
Thanks for making these updates @MontiMarco92! I'm approving this so we can merge the current state. That will make it easier to collaborate on the remaining adjustments (I'm planning on creating a new feature branch on the main remote.) It will also allow us to deploy the current state to https://next.next-drupal.org/

backlineint commented:
@MontiMarco92 I created a Draft PR with a start of the final adjustments we need: https://github.com/chapter-three/next-drupal/pull/804

The currently merged state of these docs are now available on https://next.next-drupal.org/
